### PR TITLE
Update architecture guide

### PR DESCRIPTION
**Pull request by dobooth**
## Purpose of this pull request

This PR updates the Arch guide with feedback from the Arch team.
This is mostly a reorg.
Updated TOC. 
Clarified layer names.
Moved GDPR TOC.
Fixed links.
Merged tables.

### CODE REVIEW

File: guides/v2.1/architecture/archi_perspectives/persist_layer.md [hunk 1]
```diff
@@ -14,14 +14,25 @@ If you expect to return multiple items from a database query, then you would imp
...
+However, some objects have a vast number of attributes, or they could have a set related objects that have varying numbers of attributes. In these cases, the objects are constructed using **Entity-Attribute-Value (EAV)** models.
 
 Any model that uses an EAV resource has its attributes spread out over a number of MySQL tables.
 
 The `Customer`, `Catalog` and `Order` resource models use EAV attributes.
 
+## XML Declaritive schema
+
+With Magento 2.3, we introduced Declaritive XML Schemas.
+These are XML files that are used to specify the final state of the database.
```
Comment (bdenham) [inline guides/v2.1/architecture/archi_perspectives/persist_layer.md]: Consider combining these two sentences into something like this:
With Magento 2.3, we introduced Declaritive XML Schemas, XML files used to specify the final state of the database.

File: guides/v2.1/architecture/archi_perspectives/persist_layer.md [hunk 2]
```diff
@@ -14,14 +14,25 @@ If you expect to return multiple items from a database query, then you would imp
...
+These files replace PHP update scripts that were required when upgrading a module.
+These files allow you to skip the progressive upgrade scripts and jump right to the final state of the database.
```
Comment (bdenham) [inline guides/v2.1/architecture/archi_perspectives/persist_layer.md]:
```suggestion
The files allow you to skip the progressive upgrade scripts and jump right to the final state of the database.
```

File: guides/v2.1/architecture/archi_perspectives/present_layer.md [hunk 3]
```diff
@@ -27,7 +27,21 @@ Magento themes organize both the visual aspect of your storefront and certain as
...
+For an extensive introduction to theme elements and an overview of how to extend and override the default Magento themes, see the [Frontend Developer Guide][].
+
+## GraphQL
+
+GraphQL is a data query language developed internally by Facebook in 2012 before being publicly released in 2015. Magento implements GraphQL to provide an alternative to REST and SOAP web APIs for frontend development.
+
+GraphQL allows you to define the structure of the data that you need, and the server returns only the data you request. Each GraphQL-capable module contains a declarative schema that defines the syntax for queries that the module supports, as well as the attributes that can be returned. If you run a REST call such as GET /V1/products/:sku on a simple product, the system might fetch more than 100 lines of data. If all you need is the current price, the call has returned significantly more information than you need. With GraphQL, a query against the same SKU could return just the price.
```
Comment (bdenham) [inline guides/v2.1/architecture/archi_perspectives/present_layer.md]:
```suggestion
GraphQL allows you to request only the data you need. Each GraphQL-capable module contains a declarative schema that defines the syntax for queries the module supports, as well as the attributes that can be returned. For example, a REST call, such as `GET /V1/products/:sku` on a simple product, might return more than 100 lines of data. But if all you need is the current price, you are wasting bandwidth. With GraphQL, you can query for just the price against the same SKU.
```

File: guides/v2.1/architecture/archi_perspectives/present_layer.md [hunk 4]
```diff
@@ -27,7 +27,21 @@ Magento themes organize both the visual aspect of your storefront and certain as
...
+
+More information can be found in the [GraphQL Developer Guide][]
+
+## Progressive Web Apps
```
Comment (bdenham) [inline guides/v2.1/architecture/archi_perspectives/present_layer.md]: Looks like this title needs to changed to: Progressive Web App Studio to distinguish it from the same title below.

File: guides/v2.1/architecture/archi_perspectives/tech-vision.md [hunk 5]
```diff
@@ -0,0 +1,15 @@
+---
+group: architecture-guide
+title: Magento technical vision
+---
+
+The Magento technical vision is a collection of documents that describe the desired state of the Magento platform.
+
+Each individual technical vision document relates to a set of modules that are logically grouped together. Individual documents typically describe a component's place in the system, its architecture, extension scenarios. and a list of invariants that must be preserved at all times during component refactoring or extension to ensure consistency.
```
Comment (bdenham) [inline guides/v2.1/architecture/archi_perspectives/tech-vision.md]:
```suggestion
Each individual technical vision document relates to a set of modules that are logically grouped together. Individual documents typically describe a component's place in the system, its architecture, extension scenarios, and a list of invariants that must be preserved at all times during component refactoring or extension to ensure consistency.
```

File: guides/v2.1/architecture/extensibility.md [hunk 6]
```diff
@@ -20,6 +20,24 @@ The following factors significantly affect extensibility.
...
+
+Design patterns are time-tested, widely recognized software architecture constructs. Magento product architecture incorporates many well known patterns, but Model-View-Controller (MVC) holds particular interest for extension developers.
+
+### Modularity
+
+The concept of the *module* is the heart of Magento {% glossarytooltip 55774db9-bf9d-40f3-83db-b10cc5ae3b68 %}extension{% endglossarytooltip %} development, and modular design of software components (in particular, modules, themes, and language packages) is a core architectural principle of the product. Self-contained modules of discrete code are organized by feature, thereby reducing each module's external dependencies.
+
+If a {% glossarytooltip c1e4242b-1f1a-44c3-9d72-1d5b1435e142 %}module{% endglossarytooltip %} is self-contained, then you can modify or replace it without affecting other areas of the code. This *loose coupling* of software components reduces the ripple effects throughout your code base of changing code.
```
Comment (bdenham) [inline guides/v2.1/architecture/extensibility.md]:
```suggestion
If a {% glossarytooltip c1e4242b-1f1a-44c3-9d72-1d5b1435e142 %}module{% endglossarytooltip %} is self-contained, then you can modify or replace it without affecting other areas of the code. This *loose coupling* of software components reduces the ripple effects of changing code throughout your code base.
```

File: guides/v2.1/architecture/extensibility.md [hunk 7]
```diff
@@ -32,10 +50,66 @@ Adherence to standard best practices for [PHP](https://glossary.magento.com/php)
...
 
+### Web APIs
+
+Magento or third-party services can be configured as a web {% glossarytooltip 786086f2-622b-4007-97fe-2c19e5283035 %}API{% endglossarytooltip %} (REST or SOAP) with some simple {% glossarytooltip 8c0645c5-aa6b-4a52-8266-5659a8b9d079 %}XML{% endglossarytooltip %}. You can use these services to integrate your Magento installation into third-party applications, such as CRM (Customer Relationship Management), ERP (Enterprise Resource Planning) back office systems, and {% glossarytooltip f3944faf-127e-4097-9918-a2e9c647d44f %}CMS{% endglossarytooltip %} (Content Management Systems).
+
+See [Getting Started with Magento Web APIs][] for more information.
+
+### Flexible attribute types
+
+You can enhance your storefront by adding unique attributes to the default product attributes. For example, you might need to add a new attribute to describe a product, such as texture or an industry-specific rating. You can add these attributes from the Magento Admin, and the storefront  displays them.
```
Comment (bdenham) [inline guides/v2.1/architecture/extensibility.md]:
```suggestion
You can enhance your storefront by adding unique attributes to the default product attributes. For example, you might need to add a new attribute to describe a product, such as texture or an industry-specific rating. You can add these attributes from the Magento Admin, and the storefront displays them.
```

File: guides/v2.1/architecture/extensibility.md [hunk 8]
```diff
@@ -32,10 +50,66 @@ Adherence to standard best practices for [PHP](https://glossary.magento.com/php)
...
+
+* **Custom attributes** are a subset of EAV attributes. Objects that use EAV attributes typically store values in several MySQL tables. The Customer and {% glossarytooltip 8d40d668-4996-4856-9f81-b1386cf4b14f %}Catalog{% endglossarytooltip %} modules use EAV attributes.
+
+* **Extension attributes** often use more {% glossarytooltip fd9ae55f-ccf5-480b-a7f3-bd2c80f0b2a4 %}complex data{% endglossarytooltip %} types than custom attributes. These attributes do not appear in the storefront. Extension attributes are introduced by modules.
+
+See [PHP Developer Guide][] for information about using attributes.
+
+### Service contracts, dependency injection, and dependency inversion
+
+*Service contracts* provide a new way to access public API endpoints. These PHP interfaces offer robust, stable extension points to which clients can connect.  Service contracts define the endpoints that function as a module's public API. Defining these endpoints is an essential part of adding a module.
```
Comment (bdenham) [inline guides/v2.1/architecture/extensibility.md]:
```suggestion
*Service contracts* provide a new way to access public API endpoints. These PHP interfaces offer robust, stable extension points to which clients can connect. Service contracts define the endpoints that function as a module's public API. Defining these endpoints is an essential part of adding a module.
```

File: guides/v2.1/architecture/extensibility.md [hunk 9]
```diff
@@ -32,10 +50,66 @@ Adherence to standard best practices for [PHP](https://glossary.magento.com/php)
...
+
+Service contracts are discussed throughout the Magento documentation set. See [Service layer][] for a high-level introduction. See [PHP Developer Guide][] for a more detailed discussion of service contracts and dependency injection.
+
+Magento implements *dependency injection* along with service contracts. Dependency injection provides a mechanism for changing a module's behavior without altering the client or understanding nitty-gritty details of implementation. Both dependency injection and its related concept *dependency inversion* support Magento's fundamental architectural principles of modularity and ease-of-extensibility. They strongly encourage basic coding practices that support the loose coupling of software modules.
```
Comment (bdenham) [inline guides/v2.1/architecture/extensibility.md]:
```suggestion
Magento implements *dependency injection* along with service contracts. Dependency injection provides a mechanism for changing a module's behavior without altering the client or requiring a detailed understanding of the implementation. Both dependency injection and its related concept *dependency inversion* support Magento's fundamental architectural principles of modularity and ease-of-extensibility. They strongly encourage basic coding practices that support the loose coupling of software modules.
```

File: guides/v2.1/architecture/frontend_custom_strategies.md [hunk 10]
```diff
@@ -53,14 +55,48 @@ Users of the default Magento storefront encounter examples of these form element
 
 The Magento [Admin](https://glossary.magento.com/admin) Pattern library is built on the Less preprocessor and implemented as a [module](https://glossary.magento.com/module). You can download a free, current version of this module from [Magento Marketplace](https://marketplace.magento.com/).
 
-See [Magento Admin Pattern Library]({{page.baseurl}}/pattern-library/bk-pattern.html) for more information on using this library.
+See [Magento Admin Pattern Library][] for more information on using this library.
+
+### Storefront customization levels
+
+These four levels of potential storefront customization are listed in order to increase complexity.
```
Comment (bdenham) [inline guides/v2.1/architecture/frontend_custom_strategies.md]:
```suggestion
These four levels of potential storefront customization are listed in order of increasing complexity.
```

File: guides/v2.1/architecture/frontend_custom_strategies.md [hunk 11]
```diff
@@ -53,14 +55,48 @@ Users of the default Magento storefront encounter examples of these form element
...
+
+#### Extend Magento-Provided CSS
+
+Magento supplies a default {% glossarytooltip d2093e4a-2b71-48a3-99b7-b32af7158019 %}theme{% endglossarytooltip %} and a Less-based CSS. You can substantially change a storefront using CSS only. This uncomplicated strategy might suit projects with a limited budget, or might interest developers who create different skins for a site. A small business enter this process of storefront customization by buying a third-party developed theme from Magento Marketplace to extend the default values.
```
Comment (bdenham) [inline guides/v2.1/architecture/frontend_custom_strategies.md]:
```suggestion
Magento supplies a default {% glossarytooltip d2093e4a-2b71-48a3-99b7-b32af7158019 %}theme{% endglossarytooltip %} with Less-based CSS. You can substantially change a storefront using CSS only. This uncomplicated strategy might suit projects with a limited budget, or might interest developers who create different skins for a site. A small business can use this process of storefront customization by buying a third-party developed theme from Magento Marketplace to extend the default values.
```

File: guides/v2.1/architecture/frontend_custom_strategies.md [hunk 12]
```diff
@@ -53,14 +55,48 @@ Users of the default Magento storefront encounter examples of these form element
...
+
+#### Replace PHTML template files
+
+In addition to extending the default CSS, you can generate different HTML {% glossarytooltip 8f407f13-4350-449b-9dc5-217dcf01bc42 %}markup{% endglossarytooltip %}. For example, you might need to add a missing CSS class name, or an add an extra `<div>` tag to achieve some visual effect. You might also need to tweak some {% glossarytooltip 312b4baf-15f7-4968-944e-c814d53de218 %}JavaScript{% endglossarytooltip %} to cope with different HTML markup. This change is more demanding than simply extending Magento CSS, but is still within the grasp of smaller projects and leaner teams.
```
Comment (bdenham) [inline guides/v2.1/architecture/frontend_custom_strategies.md]:
```suggestion
In addition to extending the default CSS, you can generate different HTML {% glossarytooltip 8f407f13-4350-449b-9dc5-217dcf01bc42 %}markup{% endglossarytooltip %}. For example, you might need to add a missing CSS class name, or add an extra `<div>` tag to achieve some visual effect. You might also need to tweak some {% glossarytooltip 312b4baf-15f7-4968-944e-c814d53de218 %}JavaScript{% endglossarytooltip %} to cope with different HTML markup. This change is more demanding than simply extending Magento CSS, but is still within the grasp of smaller projects and leaner teams.
```

File: guides/v2.1/architecture/frontend_custom_strategies.md [hunk 13]
```diff
@@ -53,14 +55,48 @@ Users of the default Magento storefront encounter examples of these form element
...
+
+#### Replace Magento-Provided CSS
+
+Rather than edit the default CSS provided by Magento, you might decide to replace all the default storefront CSS code with your own. This strategy avoids tying a project to the Magento-provided CSS, but puts a greater burden on project development and integration. It also allows use of different CSS tools or technologies not provided with Magento. Partners who build their own set of CSS libraries could reuse these libraries on different customer projects. (These unique CSS libraries may help differentiate a partner from others in the market.)
+
+In addition to replacing CSS files, you might need to replace small amounts of HTML and JavaScript.
+
+#### Replace Magento-Provided CSS, HTML, and JavaScript
+
+Delivering a sharply different shopping experience than the default Magento installation provides is a more substantial task. However, the tradeoff might be a more complicated experience integrating additional extensions into your installation in the future.
```
Comment (bdenham) [inline guides/v2.1/architecture/frontend_custom_strategies.md]:
```suggestion
Delivering a sharply different shopping experience than the default Magento installation provides is a more substantial task. However, the trade-off might be a more complicated experience that requires integrating additional extensions into your installation in the future.
```

File: guides/v2.1/config-guide/secy/secy.md [hunk 14]
```diff
@@ -7,7 +7,46 @@ functional_areas:
   - Setup
 ---
 
+## Enhanced password management
+
+Magento has strengthened the hashing algorithms (SHA-256) used in password management.
+
+## Improved prevention of cross-site scripting (XSS) attacks by making escaped data the default
+
+The Magento Framework has adopted conventions that regulate the escaping of data in output. These conventions include the ability to escape  output for {% glossarytooltip a2aff425-07dd-4bd6-9671-29b7edefa871 %}HTML{% endglossarytooltip %} pages (HTML, JSON, and JavaScript) and email. Where possible, escaping is transparent to client code. See [Security measures against XSS attacks][] in the {% glossarytooltip b00459e5-a793-44dd-98d5-852ab33fc344 %}Frontend{% endglossarytooltip %} Developer Guide.
```
Comment (bdenham) [inline guides/v2.1/config-guide/secy/secy.md]:
```suggestion
The Magento Framework has adopted conventions that regulate the escaping of data in output. These conventions include the ability to escape output for {% glossarytooltip a2aff425-07dd-4bd6-9671-29b7edefa871 %}HTML{% endglossarytooltip %} pages (HTML, JSON, and JavaScript) and email. Where possible, escaping is transparent to client code. See [Security measures against XSS attacks][] in the {% glossarytooltip b00459e5-a793-44dd-98d5-852ab33fc344 %}Frontend{% endglossarytooltip %} Developer Guide.
```

File: guides/v2.2/architecture/archi_perspectives/tech-vision.md [hunk 15]
```diff
@@ -0,0 +1,15 @@
+---
+group: architecture-guide
+title: Magento technical vision
+---
+
+The Magento technical vision is a collection of documents that describe the desired state of the Magento platform.
+
+Each individual technical vision document relates to a set of modules that are logically grouped together. Individual documents typically describe a component's place in the system, its architecture, extension scenarios. and a list of invariants that must be preserved at all times during component refactoring or extension to ensure consistency.
```
Comment (bdenham) [inline guides/v2.2/architecture/archi_perspectives/tech-vision.md]:
```suggestion
Each individual technical vision document relates to a set of modules that are logically grouped together. Individual documents typically describe a component's place in the system, its architecture, extension scenarios, and a list of invariants that must be preserved at all times during component refactoring or extension to ensure consistency.
```

File: guides/v2.2/architecture/frontend_custom_strategies.md [hunk 16]
```diff
@@ -52,14 +56,53 @@ Users of the default Magento storefront encounter examples of these form element
 
 The Magento [Admin](https://glossary.magento.com/admin) Pattern library is built on the Less preprocessor and implemented as a [module](https://glossary.magento.com/module). You can download a free, current version of this module from [Magento Marketplace](https://marketplace.magento.com/).
 
-See [Magento Admin Pattern Library]({{page.baseurl}}/pattern-library/bk-pattern.html) for more information on using this library.
+See [Magento Admin Pattern Library][] for more information on using this library.
+
+## Storefront customization levels
+
+These four levels of potential storefront customization are listed in order to increase complexity.
```
Comment (bdenham) [inline guides/v2.2/architecture/frontend_custom_strategies.md]:
```suggestion
These four levels of potential storefront customization are listed in order of increasing complexity.
```

File: guides/v2.2/architecture/frontend_custom_strategies.md [hunk 17]
```diff
@@ -52,14 +56,53 @@ Users of the default Magento storefront encounter examples of these form element
...
+
+In addition to replacing CSS files, you might need to replace small amounts of HTML and JavaScript.
+
+### Replace Magento-Provided CSS, HTML, and JavaScript
+
+Delivering a sharply different shopping experience than the default Magento installation provides is a more substantial task. However, the tradeoff might be a more complicated experience integrating additional extensions into your installation in the future.
+
+{:.bs-callout .bs-callout-tip}
+ Any customization of your storefront will work optimally, and provide the easiest path for later upgrades, if you follow the best practice of consistently compartmentalizing code by type. For example, keep all HTML in {% glossarytooltip ae0f1f68-c466-4189-88fd-6cd8b23c804f %}PHTML{% endglossarytooltip %} files; keep all JavaScript in JavaScript files.
```
Comment (bdenham) [inline guides/v2.2/architecture/frontend_custom_strategies.md]: See previous edits for above paragraphs.

File: guides/v2.3/architecture/archi_perspectives/tech-vision.md [hunk 18]
```diff
@@ -0,0 +1,15 @@
+---
+group: architecture-guide
+title: Magento technical vision
+---
+
+The Magento technical vision is a collection of documents that describe the desired state of the Magento platform.
+
+Each individual technical vision document relates to a set of modules that are logically grouped together. Individual documents typically describe a component's place in the system, its architecture, extension scenarios. and a list of invariants that must be preserved at all times during component refactoring or extension to ensure consistency.
```
Comment (bdenham) [inline guides/v2.3/architecture/archi_perspectives/tech-vision.md]:
```suggestion
Each individual technical vision document relates to a set of modules that are logically grouped together. Individual documents typically describe a component's place in the system, its architecture, extension scenarios, and a list of invariants that must be preserved at all times during component refactoring or extension to ensure consistency.
```

File: guides/v2.3/architecture/archi_perspectives/webapi-vision.md [hunk 19]
```diff
@@ -0,0 +1,116 @@
...
+1. Create custom resolvers.
+1. Declare a custom query/mutation and all necessary types in `schema.graphqls`.
+
+If the target module is called `MyModule`, then create the resolvers and configuration files in a new module called `MyModuleGraphQl`.
+
+See the [GraphQL documentation](https://devdocs.magento.com/guides/v2.3/graphql/index.html) for more information.
+
+#### Add a custom authentication mechanism
+
+1. Provide a custom implementation of `\Magento\Authorization\Model\UserContextInterface`, that verifies a user's identity using a  custom authentication mechanism.
```
Comment (bdenham) [inline guides/v2.3/architecture/archi_perspectives/webapi-vision.md]:
```suggestion
1. Provide a custom implementation of `\Magento\Authorization\Model\UserContextInterface` that verifies a user's identity using a  custom authentication mechanism.
```

File: guides/v2.3/architecture/archi_perspectives/webapi-vision.md [hunk 20]
```diff
@@ -0,0 +1,116 @@
...
+1. Internal server errors must be masked and never shown to the user in production mode. In developer mode, original exceptions must never be masked and should be displayed along with the related stacktrace.
+1. Pagination must be supported by all list operations.
+
+#### GraphQL
+
+1. Unlimited nesting should be supported during requests for related entities. (For example, get Order => Order Items => Products => Related Products)
+1. Field filtration must be performed with SQL queries. Do not filter on the application layer after you've fetched all possible fields.
+1. Third-party customizations must be done separately for Service Contracts and for GraphQL
+1. For modularity purposes, GraphQL configuration must be declared in a separate module. For example, to expose GraphQL for the module `MyModule`, you must create the `schema.graphqls` file in the `MyModuleGraphQl` module.
+1. GraphQL is primarily designed for store-front one-page apps and mobile applications. It supports token and cookie authentication, as well as guest access to public queries
```
Comment (bdenham) [inline guides/v2.3/architecture/archi_perspectives/webapi-vision.md]:
```suggestion
1. GraphQL is primarily designed for store-front, one-page apps and mobile applications. It supports token and cookie authentication, as well as guest access to public queries
```

File: guides/v2.3/architecture/archi_perspectives/webapi-vision.md [hunk 21]
```diff
@@ -0,0 +1,116 @@
...
+1. The Store code should be passed via headers.
+
+#### REST
+
+1. The resource URL should be versioned (for example: V1). The version must be specified in the following format: `V\d.+`
+1. Resource names in a URL should be in plural form (for example: products, carts)
+1. ID parameters for operations on entities should be part of the resource URL (for example: /V1/products/**:sku** )
+1. POST should be used to create an entity.
+   PUT should be used to update an entity.
+   PATCH should be used to patch (update only selected fields) an entity(s) update.
```
Comment (bdenham) [inline guides/v2.3/architecture/archi_perspectives/webapi-vision.md]:
```suggestion
   PATCH should be used to patch (update only selected fields) an update to an entity or entities.
```

File: guides/v2.3/architecture/archi_perspectives/webapi-vision.md [hunk 22]
```diff
@@ -0,0 +1,116 @@
...
+1. REST is designed for system integrations, mobile app integrations, as well as for one-page apps. It supports tokens, cookies and OAuth 1.0 with token exchange. It also supports guest access to public resources.
```
Comment (bdenham) [inline guides/v2.3/architecture/archi_perspectives/webapi-vision.md]:
```suggestion
1. REST is designed for system integrations, mobile app integrations, as well as for one-page apps. It supports tokens, cookies, and OAuth 1.0 with token exchange. It also supports guest access to public resources.
```

File: guides/v2.3/architecture/archi_perspectives/webapi-vision.md [hunk 23]
```diff
@@ -0,0 +1,116 @@
...
+1. Responses must return responses with standard [HTTP status codes](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes).
+1. The store code must be passed via URL. For example `GET /rest/frenchStoreView/V1/products`. Persistence operations that should be performed for all stores at once, should have 'all' store code in the URL.
```
Comment (bdenham) [inline guides/v2.3/architecture/archi_perspectives/webapi-vision.md]:
```suggestion
1. The store code must be passed with a URL. For example `GET /rest/frenchStoreView/V1/products`. Persistence operations that should be performed for all stores at once should have 'all' store code in the URL.
```

**Comment by dobooth** [comment]
running tests

**Comment by dobooth** [comment]
running tests

**Comment by dobooth** [comment]
running tests

**Pull request merged**